### PR TITLE
deserializer handler testing 

### PR DESCRIPTION
**Pull request by erk-dik**
Tests covering a subsets of the available deserializar handlers.

### CODE REVIEW

File: test/LionWeb-CSharp-Test/tests/serialization/DeserializationTests.cs [hunk 1]
```diff
@@ -269,4 +280,773 @@ public void deserializeUnsetRequiredContainment()
...
+
+        IDeserializer deserializer = new DeserializerBuilder()
+            .WithHandler(new DeserializerExceptionHandler())
+            .WithLanguage(ShapesLanguage.Instance)
+            .Build();
+
+        Assert.ThrowsException<UnsupportedClassifierException>(() => deserializer.Deserialize(serializationChunk));
+    }
+
+    private class UnknownClassifierHandler(Func<Classifier?> incrementer) : IDeserializerHandler
```
Comment (dslmeinte) [inline test/LionWeb-CSharp-Test/tests/serialization/DeserializationTests.cs:315]: Would it be an idea to have an abstract `NotImplementedDeserializerHandler` which has every method throw an `NotImplementedException`. Then the `UnknownClassifierHandler` could extend that and just implement `UnknownClassifier`, etc.

File: test/LionWeb-CSharp-Test/tests/serialization/DeserializationTests.cs [hunk 2]
```diff
@@ -269,4 +280,773 @@ public void deserializeUnsetRequiredContainment()
...
+
+        IDeserializer deserializer = new DeserializerBuilder()
+            .WithHandler(new DeserializerExceptionHandler())
+            .WithLanguage(ShapesLanguage.Instance)
+            .Build();
+
+        Assert.ThrowsException<UnknownFeatureException>(() => deserializer.Deserialize(serializationChunk));
+    }
+
+    private class UnknownFeatureHandler(Func<Feature?> incrementer) : IDeserializerHandler
```
Comment (dslmeinte) [inline test/LionWeb-CSharp-Test/tests/serialization/DeserializationTests.cs:442]: The `NotImplementedDeserializerHandler` proposed above would also make things more readable here.

File: test/LionWeb-CSharp-Test/tests/serialization/DeserializationTests.cs [hunk 3]
```diff
@@ -269,4 +280,773 @@ public void deserializeUnsetRequiredContainment()
...
+                    Containments = [],
+                    References = [],
+                    Annotations = ["lizard"],
+                }
+            ]
+        };
+
+        var count = 0;
+
+        IDeserializer deserializer = new DeserializerBuilder()
```
Comment (dslmeinte) [inline test/LionWeb-CSharp-Test/tests/serialization/DeserializationTests.cs:1032]: I'm guessing the construction of the `IDeserializer` instances could be DRY'd a bit.

**Review by dslmeinte — CHANGES_REQUESTED** [review]
(draft)

File: test/LionWeb-CSharp-Test/tests/serialization/DeserializationTests.cs (hunk 2 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/DeserializationTests.cs]: This one and similar ones now extend and specialize NotImplementedDeserializerHandler .

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnknownClassifierTests.cs [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+namespace LionWeb_CSharp_Test.tests.serialization.deserialization;
+
+using Examples.Shapes.M2;
+using LionWeb.Core;
+using LionWeb.Core.M1;
+using LionWeb.Core.M3;
+using LionWeb.Core.Serialization;
+
+/// <summary>
+/// <see cref="IDeserializerHandler.UnknownClassifier"/>
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnknownClassifierTests.cs:27]: I don't think this needs / should be inside the `<summary>` tag

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnknownClassifierTests.cs [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+            .WithHandler(deserializerHealingHandler)
+            .WithLanguage(ShapesLanguage.Instance)
+            .Build();
+
+        List<IReadableNode> deserializedNodes = deserializer.Deserialize(serializationChunk);
+        Assert.AreEqual(0, deserializedNodes.Count);
+    }
+
+    /// <summary>
+    /// Heals by replacing unknown classifier with a known classifier in the language
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnknownClassifierTests.cs:73]: which language?

File: src/LionWeb-CSharp/core/M1/IDeserializerHandler.cs [hunk 6]
```diff
@@ -38,6 +38,7 @@ public interface IDeserializerHandler
     /// <param name="node">Currently deserialized node with id <paramref name="nodeId"/>.</param>
     /// <returns>Replacement node id to use for <paramref name="node"/>, or <c>null</c> to skip <paramref name="node"/>.</returns>
     /// <remarks>For both <paramref name="existingNode"/> and <paramref name="node"/>, only node id and properties are populated -- no other features.</remarks>
+    /// <remarks>if returned replacement node id is not unique, deserializer keeps looking for an unique node id, this may lead to an infinite loop </remarks>
```
Comment (enikao) [inline src/LionWeb-CSharp/core/M1/IDeserializerHandler.cs:41]: deserializer keeps calling this method, that might be the infinite loop

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs [hunk 7]
```diff
@@ -0,0 +1,872 @@
...
+
+    /// <summary>
+    /// Test case:
+    /// Direct circular dependency: A has B as child, B has A as child 
+    /// Healing:
+    /// Breaks containment link from A to B. As a result, A does not have B as a child and B has A as a child  
+    /// </summary>
+    /// <remarks>Instead of breaking containment link from A to B, a reference link can be created from A to B</remarks>
+    [TestMethod]
+    public void direct_circular_containment_heals_case_2()
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs:166]: explain in docs what's the difference to the test above

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs [hunk 8]
```diff
@@ -0,0 +1,872 @@
...
+
+        INode a = b.Children().First();
+        Assert.AreEqual("A", a.GetId());
+        Assert.AreSame(b, a.GetParent());
+        Assert.IsFalse(a.Children().Any());
+    }
+
+    /// <summary>
+    /// Test case:
+    /// Direct circular dependency: A has B, C as children. B has C as a child. C has A as a child.
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs:237]: `B` has `D` as child I think?

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs [hunk 9]
```diff
@@ -0,0 +1,872 @@
...
+
+        INode d = b.Children().First();
+        Assert.AreEqual("D", d.GetId());
+        Assert.AreSame(b, d.GetParent());
+        Assert.IsFalse(d.Children().Any());
+    }
+
+    /// <summary>
+    /// Test case:
+    /// Direct circular dependency: A has B, C as children. B has C as a child. C has A as a child.
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs:350]: `B` has `D` as child

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs [hunk 10]
```diff
@@ -0,0 +1,872 @@
...
+    /// Healing:
+    /// Breaks containment link from A to C.
+    /// </summary>
+    [TestMethod]
+    public void direct_circular_containment_multiple_children_heals_case_2()
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs:355]: explain difference to previous test

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs [hunk 11]
```diff
@@ -0,0 +1,872 @@
...
+    }
+
+    /// <summary>
+    /// Test case:
+    /// Indirect circular dependency: A has B as a child, B has C as a child, C has A as a child.
+    /// Healing:
+    /// Breaks containment link from B to C.
+    /// </summary>
+    [TestMethod]
+    public void indirect_circular_containment_case_2()
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs:566]: explain difference to previous test

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs [hunk 12]
```diff
@@ -0,0 +1,872 @@
...
+                    Containments =
+                    [
+                        new SerializedContainment()
+                        {
+                            Containment = new MetaPointer("key-Shapes", "1", "key-default-group"), Children = []
+                        }
+                    ],
+                    References = [],
+                    Annotations = ["A"],
+                    Parent = null
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs:690]: Why not `A`?

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/DuplicateContainmentTests.cs [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+    private class DeserializerHealingHandler(Func<IWritableNode, IWritableNode, IReadableNode, bool> heal)
+        : DeserializerExceptionHandler
+    {
+        public override bool DuplicateContainment(IWritableNode containedNode, IWritableNode newParent,
+            IReadableNode existingParent) =>
+            heal(containedNode, newParent, existingParent);
+    }
+
+    [TestMethod]
+    public void duplicate_containment_keeps_existing_parent()
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/DuplicateContainmentTests.cs:40]: this class is already named `DuplicateContainment`, no need to repeat in method name

Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/DuplicateContainmentTests.cs:40]: Both parents are existing parents. It again depends on the processing order which one comes first (similar to circular containment).

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidAnnotationTests.cs [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+        IDeserializer deserializer = new DeserializerBuilder()
+            .WithHandler(deserializerHealingHandler)
+            .WithLanguage(ShapesLanguage.Instance)
+            .Build();
+
+        List<IReadableNode> deserializedNodes = deserializer.Deserialize(serializationChunk);
+        Assert.AreEqual(2, deserializedNodes.Count);
+        Assert.AreEqual(billOfMaterials, deserializedNodes.OfType<Circle>().FirstOrDefault()?.GetAnnotations()[0]);
+    }
+}
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidAnnotationTests.cs:129]: Add another test: Heal with another invalid annotation

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidFeatureTests.cs [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+{
+    private class DeserializerHealingHandler(Func<CompressedMetaPointer, Classifier, IWritableNode, Feature?> heal)
+        : DeserializerExceptionHandler
+    {
+        public override Feature? InvalidFeature<TFeature>(CompressedMetaPointer feature, Classifier classifier,
+            IWritableNode node) => heal(feature, classifier, node);
+    }
+
+    [TestMethod]
+    public void invalid_feature_does_not_heal()
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidFeatureTests.cs:40]: maybe add docs: existing feature in wrong classifier

Comment (enikao) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidFeatureTests.cs:40]: or adjust method name

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/SkipDeserializingDependentNodeTests.cs [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+public class SkipDeserializingDependentNodeTests
+{
+    private class DeserializerHealingHandler(Func<CompressedId, bool> heal) : DeserializerExceptionHandler
+    {
+        public override bool SkipDeserializingDependentNode(CompressedId id) => heal(id);
+    }
+
+
+    [TestMethod]
+    [Ignore(
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/SkipDeserializingDependentNodeTests.cs:39]: didn't we do this by now?

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnknownFeatureTests.cs [hunk 17]
```diff
@@ -0,0 +1,351 @@
...
+        IDeserializer deserializer = new DeserializerBuilder()
+            .WithHandler(deserializerHealingHandler)
+            .WithLanguage(LibraryLanguage.Instance)
+            .Build();
+
+        List<IReadableNode> deserializedNodes = deserializer.Deserialize(serializationChunk);
+        Assert.AreEqual(2, deserializedNodes.Count);
+        Assert.AreEqual(1,
+            deserializedNodes.OfType<Book>().FirstOrDefault()?.CollectAllSetFeatures().OfType<Reference>().Count());
+    }
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnknownFeatureTests.cs:350]: add tests that heal to invalid feature

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/DuplicateNodeIdTests.cs [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+
+        Assert.IsNotNull(last.Offset);
+        Assert.AreEqual("ChildSecond", last.Offset.GetId());
+
+        Assert.IsNotNull(last.Source);
+        Assert.AreEqual("RefSecond", last.Source.GetId());
+
+        Assert.AreEqual(2, nodes.OfType<Circle>().Count());
+        Assert.AreEqual(0, nodes.OfType<Coord>().Count());
+    }
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/DuplicateNodeIdTests.cs:354]: add test that heals the first time to still duplicate node, only works the second time

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidFeatureTests.cs [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+            .WithHandler(deserializerHealingHandler)
+            .WithLanguage(ShapesLanguage.Instance)
+            .WithUncompressedIds(true)
+            .Build();
+
+        List<IReadableNode> deserializedNodes = deserializer.Deserialize(serializationChunk);
+        Assert.AreEqual(1,
+            deserializedNodes.OfType<OffsetDuplicate>().FirstOrDefault()?.CollectAllSetFeatures().OfType<Reference>().Count());
+        Assert.AreEqual(2, deserializedNodes.Count);
+    }
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidFeatureTests.cs:207]: add tests that heal to invalid feature

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidLinkValueTests.cs [hunk 20]
```diff
@@ -0,0 +1,670 @@
...
+        IDeserializer deserializer = new DeserializerBuilder()
+            .WithHandler(deserializerHealingHandler)
+            .WithLanguage(TinyRefLangLanguage.Instance)
+            .Build();
+
+        List<IReadableNode> deserializedNodes = deserializer.Deserialize(serializationChunk);
+        Assert.AreEqual(3, deserializedNodes.Count);
+        MyConcept myConcept = deserializedNodes.OfType<MyConcept>().First();
+        Assert.IsTrue(myConcept.CollectAllSetFeatures().Contains(TinyRefLangLanguage.Instance.MyConcept_singularRef));
+    }
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidLinkValueTests.cs:669]: add tests that heal to invalid value

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnknownEnumerationLiteralTests.cs [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+        IDeserializer deserializer = new DeserializerBuilder()
+            .WithHandler(deserializerHealingHandler)
+            .WithLanguage(WithEnumLanguage.Instance)
+            .Build();
+
+        List<IReadableNode> deserializedNodes = deserializer.Deserialize(serializationChunk);
+        Assert.IsTrue(deserializedNodes.OfType<EnumHolder>().FirstOrDefault()?.CollectAllSetFeatures()
+            .OfType<Property>().Contains(WithEnumLanguage.Instance.EnumHolder_enumValue));
+        Assert.AreEqual(myEnumLiteral, deserializedNodes.OfType<EnumHolder>().FirstOrDefault()?.EnumValue);
+    }
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnknownEnumerationLiteralTests.cs:126]: add test that heals to invalid value

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnresolvableAnnotationTests.cs [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+            .WithHandler(deserializerHealingHandler)
+            .WithLanguage(ShapesLanguage.Instance)
+            .Build();
+
+        List<IReadableNode> deserializedNodes = deserializer.Deserialize(serializationChunk);
+        Assert.AreEqual(1, deserializedNodes.Count);
+        INode? annotation = deserializedNodes.OfType<Circle>().FirstOrDefault()?.GetAnnotations()[0];
+        Assert.AreSame(documentation, annotation);
+        Assert.AreSame(documentation.GetId(), annotation?.GetId());
+    }
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnresolvableAnnotationTests.cs:114]: add test that heals to invalid annotation

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnresolvableChildTests.cs [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+
+        var deserializerHealingHandler = new DeserializerHealingHandler((id, feature, arg3) => circle);
+        IDeserializer deserializer = new DeserializerBuilder()
+            .WithHandler(deserializerHealingHandler)
+            .WithLanguage(ShapesLanguage.Instance)
+            .Build();
+
+        List<IReadableNode> deserializedNodes = deserializer.Deserialize(serializationChunk);
+        Assert.AreSame(circle, deserializedNodes.OfType<Geometry>().First().Shapes[0]);
+    }
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnresolvableChildTests.cs:121]: add test that heals to invalid child

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnresolvableReferenceTargetTests.cs [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+            .WithLanguage(ShapesLanguage.Instance)
+            .Build();
+
+        List<IReadableNode> deserializedNodes = deserializer.Deserialize(serializationChunk);
+        Assert.AreEqual(1, deserializedNodes.Count);
+        
+        var source = deserializedNodes.OfType<OffsetDuplicate>().FirstOrDefault()?.Source;
+        Assert.AreSame(circle, source);
+        Assert.AreSame(circle.GetId(), source?.GetId());
+    }
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnresolvableReferenceTargetTests.cs:139]: add test that heals to invalid target

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnresolvableReferenceTargetTests.cs [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+}
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnresolvableReferenceTargetTests.cs:140]: add test that heals based on the correct ResolveInfo

**Review by enikao — COMMENTED** [review]
very thorough work! Some proposals for testing failure cases.

File: src/LionWeb-CSharp/core/M1/IDeserializerHandler.cs (hunk 6 above)
Comment (erk-dik) [reply src/LionWeb-CSharp/core/M1/IDeserializerHandler.cs:41]: correct, fixed.

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs (hunk 7 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs:166]: added explanation in xdoc.

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs (hunk 8 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs:237]: corected.

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs (hunk 9 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs:350]: corrected/.

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs (hunk 10 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs:355]: added.

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs (hunk 11 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/CircularContainmentTests.cs:566]: added.

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/DuplicateContainmentTests.cs (hunk 13 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/DuplicateContainmentTests.cs:40]: got your point, I use now the wording "current parent" instead od existing parent.

Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/DuplicateContainmentTests.cs:40]: actually, i see no harm with repating here, also by reading method names,  i can differentiate tests related to containment and annotation.

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidFeatureTests.cs (hunk 15 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidFeatureTests.cs:40]: actually here we use existing feature in wrong feature type (key-r is property, we use it in reference), docs added.

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidAnnotationTests.cs (hunk 14 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidAnnotationTests.cs:129]: added

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnknownFeatureTests.cs (hunk 17 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnknownFeatureTests.cs:398]: added.

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidFeatureTests.cs (hunk 19 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidFeatureTests.cs:211]: added.

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidLinkValueTests.cs (hunk 20 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidLinkValueTests.cs:669]: added.

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidPropertyValueTests.cs [hunk 26]
```diff
@@ -0,0 +1,123 @@
...
+        IDeserializer deserializer = new DeserializerBuilder()
+            .WithHandler(deserializerHealingHandler)
+            .WithLanguage(ShapesLanguage.Instance)
+            .Build();
+
+        List<IReadableNode> deserializedNodes = deserializer.Deserialize(serializationChunk);
+        Assert.AreEqual(1, deserializedNodes.Count);
+        Assert.IsTrue(deserializedNodes.OfType<Coord>().FirstOrDefault()?.CollectAllSetFeatures().OfType<Property>()
+            .Contains(ShapesLanguage.Instance.Coord_x));
+    }
```
[the remark this answers is not in the document]
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/InvalidPropertyValueTests.cs:122]: added.

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnknownEnumerationLiteralTests.cs (hunk 21 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnknownEnumerationLiteralTests.cs:126]: added.

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnresolvableAnnotationTests.cs (hunk 22 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnresolvableAnnotationTests.cs:114]: added.

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnresolvableChildTests.cs (hunk 23 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnresolvableChildTests.cs:121]: added.

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnresolvableReferenceTargetTests.cs (hunk 24 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnresolvableReferenceTargetTests.cs:139]: added.

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnresolvableReferenceTargetTests.cs (hunk 25 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/UnresolvableReferenceTargetTests.cs:140]: added

File: test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/DuplicateNodeIdTests.cs (hunk 18 above)
Comment (erk-dik) [reply test/LionWeb-CSharp-Test/tests/serialization/deserialization/customHandler/DuplicateNodeIdTests.cs:528]: added

**Comment by erk-dik** [comment]
> Very general idea, not sure whether you want to implement it due to timing:
> 
> 1. The tests get very long, maybe we can build our own small internal "mini-DSL" to create serialized nodes? Could look similar as the language building mini-DSL in `ShapesDefinition.cs`
> 2. Currently we're checking the outcome of deserialization by a big bunch of asserts. Maybe we could just create the expected node model and use `Comparer` to assert they're equal?

I will implement those in seperate branches.

**Comment by enikao** [comment]
for all Link-based tests: Please add a test that we can handle multiple entries.

**Comment by enikao** [comment]
Depends on https://github.com/gragra33/Utf8JsonAsyncStreamReader/issues/3

**Comment by enikao** [comment]
See also https://github.com/LionWeb-io/lionweb-csharp/issues/34

**Review by enikao — APPROVED** [review]
